### ISSUE TITLE
keytab and SPN

### ISSUE THREAD

**Issue by keith6014** [jcmturner/gokrb5#309]
Great library! So great I think I am going to run my web application directly in Go instead of HAproxy!

In a windows environment, I registered the SPN with service account "foo", but I am clueless on how to setup the keytab.

My question are 1) what is the preferred tool to use to create keytab? 2) In my keytab file, what do I need? I am guessing an entry with "foo" credentials and another entry with SPN credentials. 3) How do you select the encoding type? Is it possible to determine it from KDC/Domain controller? I think if I enable tracing I can get that information. 

Do you have a full end to end example of it?

**Comment by bwmarrin** [comment]
I'm in the same boat here, I've in the past used Apache to handle Kerberos auth for web applications in Active Directory environments.

I'd really love to switch to this library and remove the dependency on Apache to handle the Kerberos auth. 

I've build a half working example from the examples in this repo but I'm not getting things working.

Is there a ELI5 howto guide on what needs done to build a web service using this? 

@keith6014, rather it's the right way or not, I generated my keytab by logging into the domain controller and running the below.  It's how I've done it in the past when creating keytab for Apache.

```
ktpass -princ HTTP/host.domain.tld@DOMAIN.TLD -mapuser user@DOMAIN.TLD -pass SomePasswordHere -crypto AES256-SHA1 -ptype KRB5_NT_PRINCIPAL -out c:\nameme.keytab
```

You'll need to adjust host.domain.tld and DOMAIN.TLD values and such.

**Comment by keith6014** [comment]
@bwmarrin 
thannks. 
How did you determine your crypto type, AES256-SHA1?
Also, you have any sample golang code which uses SPN? I haven't tried so far.

**Comment by bwmarrin** [comment]
I ended up switching from AES256-SHA1 to "All" because I had some issues later on.  Using "All" just embeds all crypto options within the key, so whatever form the browser sends you'll have a corresponding match in the SPN file.

```
ktpass -princ HTTP/host.domain.tld@DOMAIN.TLD -mapuser user@DOMAIN.TLD -pass SuperSecretPassphrase -ptype KRB5_NT_PRINCIPAL -crypto All -out c:\vbp.keytab
```
I build a test starting with https://github.com/jcmturner/gokrb5/blob/master/examples/httpServer.go

I just swapped out the section where it uses the test keytab file and put in my keytab.  I ran into another error or two and ended up having to modify this library and comment out this line https://github.com/jcmturner/gokrb5/blob/master/spnego/negotiationToken.go#L174 in order to get things working.

See,
https://github.com/jcmturner/gokrb5/issues/310

I've deleted my test example I made as I went on to incorporate it into the larger app I was writing but if the above hints don't get you in the right direction let me know and I can put together a simple example again.

**Comment by jcmturner** [comment]
@keith6014 I use ktutil to create keytabs. There is good documentation on line for it. With regards to the values to use these are implementation specific for your kerberos environment. Therefore you will need to speak with the people that run the kerberos servers to know what will work for you. You questions are not specific to the gokrb5 library so I’m going to close this issue. There are lots of good resources online to cover the principles of Kerberos.
